src/Algebra/Linear/Properties/VectorSpace.agda

```diff
@@ -0,0 +1,191 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Properties of linear algebraic structures.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra
+  using (CommutativeRing; Congruent₁; Congruent₂; Op₁; Op₂)
+open import Algebra.Module        using (Module)
+open import Algebra.Linear.Bundles
+open import Level                 using (Level; _⊔_; suc)
+
+module Algebra.Linear.Properties.VectorSpace
+  {r ℓr m ℓm : Level}
+  {ring      : CommutativeRing r ℓr}
+  {mod       : Module ring m ℓm}
+  (vs        : VectorSpace mod)
+  where
+
+open import Algebra.Module.Construct.TensorUnit using (⟨module⟩)
+open import Algebra.Module.Morphism.Bundles
+import      Algebra.Module.Morphism.ModuleHomomorphism as ModHomo
+import      Algebra.Module.Morphism.Structures as MorphismStructures
+open import Axiom.DoubleNegationElimination
+open import Data.List
+open import Data.Product
+open import Function
+open import Relation.Binary
+import      Relation.Binary.PropositionalEquality      as Eq
+open import Relation.Binary.Reasoning.MultiSetoid
+
+open VectorSpace                        vs
+open MorphismStructures.ModuleMorphisms mod ⟨module⟩
+
+import Function.Relation.Binary.Equality as ExtEq
+open ExtEq setoid
+-- import Function.Relation.Binary.Equality2 as ExtEq
+-- open ExtEq ≈ᴹ-setoid setoid
+
+------------------------------------------------------------------------
+-- Congruency of `IsBasis` helper functions.
+vscale-cong : ∀ f → Congruent _≈ᴹ_ _≈_ f → Congruent _≈ᴹ_ _≈ᴹ_ (vscale f)
+vscale-cong f f-cong {x} {y} x≈y = begin⟨ ≈ᴹ-setoid ⟩
+  vscale f x ≡⟨⟩
+  f x *ₗ x   ≈⟨ *ₗ-congʳ (f-cong x≈y) ⟩
+  f y *ₗ x   ≈⟨ *ₗ-congˡ x≈y ⟩
+  f y *ₗ y   ≡⟨⟩
+  vscale f y ∎
+
+vgen-cong : ∀ {f₁ f₂} → ∀ xs → f₁ ≗ f₂ → vgen f₁ xs ≈ᴹ vgen f₂ xs
+vgen-cong {f₁} {f₂} []       f₁≗f₂ = Setoid.reflexive ≈ᴹ-setoid Eq.refl
+vgen-cong {f₁} {f₂} (x ∷ xs) f₁≗f₂ = begin⟨ ≈ᴹ-setoid ⟩
+  f₁ x *ₗ x +ᴹ vgen f₁ xs ≈⟨ +ᴹ-congʳ (*ₗ-congʳ (f₁≗f₂ x)) ⟩
+  f₂ x *ₗ x +ᴹ vgen f₁ xs ≈⟨ +ᴹ-congˡ (vgen-cong xs f₁≗f₂) ⟩
+  f₂ x *ₗ x +ᴹ vgen f₂ xs ∎
+
+------------------------------------------------------------------------
+-- Some consequences of `VectorSpace` inner product properties.
+v∙g[x]+y-cong₂ : {g : V → V} {v x w : V} {y z : S} →
+                 Congruent _≈ᴹ_ _≈ᴹ_ g → x ≈ᴹ w → y ≈ z →
```

Comment:
Could the `Congruent` proof go before the `v x w y z` declarations? That allows the function to be partially applied more easily. Similar comment with several others below, e.g. `foldr-homo-∙`

Reply:
Changed as requested in next push.